### PR TITLE
pallet-revive: Add stateful address mapping

### LINKED ISSUE #5576: Add a mapping where Account32 wallets can register a 20 -> 32 mapping.
So far you should not call a contract with an AccountId32 because there is no way to map back from its mapped AccountId20. For this to work we need a mapping that helps mapping back from AccountId20 to AccountId32.

### PR DESCRIPTION
**Pull request by athei**
Fixes #5576

This allows contracts to be used with an AccountId32 through normal extrinsics and not only through the eth compat layer. It works by adding a new extrinsic `map_account` that lets people register their AccountId32.

### CODE REVIEW

**Comment by athei** [comment]
bot bench substrate-pallet --pallet=pallet_revive

**Comment by athei** [comment]
bot bench substrate-pallet --pallet=pallet_revive

**Comment by athei** [comment]
bot bench substrate-pallet --pallet=pallet_revive

**Comment by athei** [comment]
bot clean

**Review by xermicus — APPROVED** [review]
LGTM!

**Review by pgherveou — APPROVED** [review]

File: substrate/frame/revive/src/lib.rs [hunk 1]
```diff
@@ -963,6 +978,51 @@ pub mod pallet {
...
+		pub fn unmap_account(origin: OriginFor<T>) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			T::AddressMapper::unmap(&origin)
+		}
+
+		/// Dispatch an `call` with the origin set to the callers fallback address.
+		///
+		/// Every `AccountId32` can control its corresponding fallback account. The fallback account
+		/// is the `AccountId20` with the last 12 bytes set to `0xEE`. This is essentially a
+		/// recovery function in case an `AccountId20` was used without creating a mapping first.
```
[the remark this answers is not in the document]
Comment (athei) [reply substrate/frame/revive/src/lib.rs:1039]: You can't use it as origin. But nobody is stopping you from passing your derived `AccountId20` as an argument to a contract without creating a mapping first. Wallets should prevent that by not showing the derived address without a mapping. But we can't prevent it on a runtime level.

File: substrate/frame/revive/src/address.rs [hunk 2]
```diff
@@ -17,61 +17,172 @@
...
+	fn is_mapped(account_id: &T::AccountId) -> bool {
+		let account_bytes: &[u8; 32] = account_id.as_ref();
+		&account_bytes[20..] == &[0xEE; 12] ||
+			<AddressSuffix<T>>::contains_key(Self::to_address(account_id))
+	}
+}
+
+impl<T> AddressMapper<T> for H160Mapper<T>
+where
+	T: Config<AccountId = H160>,
```
[the remark this answers is not in the document]
Comment (athei) [reply substrate/frame/revive/src/address.rs:162]: Frontier is using `H160` aren't they? None the less I applied your suggestion to make it more compatible.

**Pull request merged**